File: snippets/screens/swiftui/tags_screen/1/tags_screen.swift
import SwiftUI

struct Tag {
    let id: String
    let name: String
    let color: Color
    let count: Int
    let category: String
    let isPopular: Bool
}

struct TagsScreen: View {
    let selectedTags: [String]?

    @State private var allTags: [Tag] = []
    @State private var filteredTags: [Tag] = []
    @State private var selectedTagIds = Set<String>()
    @State private var selectedCategory = "all"
    @State private var sortBy = "popularity"
    @State private var searchText = ""
    @State private var isLoading = true
    @State private var currentTab = 0

    private let categories = [
        "all", "technology", "design", "business", "development",
        "marketing", "science", "education", "health", "lifestyle"
    ]

    @Environment(\.dismiss) private var dismiss

    init(selectedTags: [String]? = nil) {
        self.selectedTags = selectedTags
    }

    var body: some View {
        NavigationView {
            VStack(spacing: 0) {
                // Search Bar
                HStack {
                    TextField("Search tags...", text: $searchText)
                        .textFieldStyle(.roundedBorder)
                        .onChange(of: searchText) { _, _ in
                            filterTags()
                        }

                    if !searchText.isEmpty {
                        Button {
                            searchText = ""
                            filterTags()
                        } label: {
                            Image(systemName: "xmark.circle.fill")
                                .foregroundColor(.gray)
                        }
                    }
                }
                .padding(.horizontal)

                // Filters and Sort
                HStack(spacing: 12) {
                    // Category Filter
                    Picker("Category", selection: $selectedCategory) {
                        ForEach(categories, id: \.self) { category in
                            Text(category == "all" ? "All Categories" : category.capitalized)
                                .tag(category)
                        }
                    }
                    .pickerStyle(.menu)
                    .onChange(of: selectedCategory) { _, _ in
                        filterTags()
                    }

                    Spacer()

                    // Sort Picker
                    Picker("Sort", selection: $sortBy) {
                        Text("Most Popular").tag("popularity")
                        Text("A-Z").tag("alphabetical")
                        Text("Recently Added").tag("recent")
                    }
                    .pickerStyle(.menu)
                    .onChange(of: sortBy) { _, _ in
                        sortTags()
                    }
                }
                .padding(.horizontal)
                .padding(.vertical, 8)

                // Tab View
                TabView(selection: $currentTab) {
                    browseTab
                        .tabItem {
                            Image(systemName: "list.bullet")
                            Text("Browse")
                        }
                        .tag(0)

                    popularTab
                        .tabItem {
                            Image(systemName: "chart.line.uptrend.xyaxis")
                            Text("Popular")
                        }
                        .tag(1)

                    myTagsTab
                        .tabItem {
                            Image(systemName: "bookmark.fill")
                            Text("My Tags")
                        }
                        .tag(2)
                }
            }
            .navigationTitle("Tags (\(selectedTagIds.count) selected)")
            .navigationBarTitleDisplayMode(.large)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    if !selectedTagIds.isEmpty {
                        Button("Apply") {
                            applyTags()
                        }
                        .fontWeight(.semibold)
                    }
                }
            }
        }
        .onAppear {
            loadTags()
        }
        .safeAreaInset(edge: .bottom) {
            if !selectedTagIds.isEmpty {
                bottomActionBar
            }
        }
    }

    private var browseTab: some View {
        Group {
            if isLoading {
                VStack {
                    Spacer()
                    ProgressView()
                    Spacer()
                }
            } else if filteredTags.isEmpty {
                emptyStateView
            } else {
                ScrollView {
                    LazyVGrid(columns: Array(repeating: GridItem(.flexible(), spacing: 12), count: 2), spacing: 12) {
                        ForEach(filteredTags, id: \.id) { tag in
                            tagCard(tag)
                        }
                    }
                    .padding()
                }
            }
        }
    }

    private var popularTab: some View {
        let popularTags = allTags.filter { $0.isPopular }.sorted { $0.count > $1.count }

        return List {
            ForEach(popularTags.indices, id: \.self) { index in
                let tag = popularTags[index]
                tagListItem(tag, rank: index + 1)
            }
        }
        .listStyle(.plain)
    }

    private var myTagsTab: some View {
        let myTags = allTags.filter { selectedTagIds.contains($0.id) }

        return Group {
            if myTags.isEmpty {
                VStack(spacing: 16) {
                    Spacer()

                    Image(systemName: "bookmark")
                        .font(.system(size: 64))
                        .foregroundColor(.gray)

                    Text("No tags selected")
                        .font(.title2)
                        .fontWeight(.bold)

                    Text("Browse tags and add them to your collection")
                        .font(.body)
                        .foregroundColor(.secondary)
                        .multilineTextAlignment(.center)

                    Button("Browse Tags") {
                        currentTab = 0
                    }
                    .buttonStyle(.borderedProminent)

                    Spacer()
                }
                .padding()
            } else {
                ScrollView {
                    LazyVStack(spacing: 12) {
                        ForEach(myTags, id: \.id) { tag in
                            selectedTagChip(tag)
                        }
                    }
                    .padding()
                }
            }
        }
    }

    private var emptyStateView: some View {
        VStack(spacing: 16) {
            Spacer()

            Image(systemName: "magnifyingglass.circle")
                .font(.system(size: 64))
                .foregroundColor(.gray)

            Text("No tags found")
                .font(.title2)
                .fontWeight(.bold)

            Text("Try adjusting your search or filters")
                .font(.body)
                .foregroundColor(.secondary)

            Spacer()
        }
    }

    private var bottomActionBar: some View {
        VStack(spacing: 0) {
            Divider()

            HStack(spacing: 16) {
                Button("Clear All") {
                    selectedTagIds.removeAll()
                }
                .frame(maxWidth: .infinity)
                .padding()
                .background(Color(.systemGray6))
                .foregroundColor(.primary)
                .cornerRadius(10)

                Button("Apply \(selectedTagIds.count) Tags") {
                    applyTags()
                }
                .frame(maxWidth: .infinity)
                .padding()
                .background(Color.accentColor)
                .foregroundColor(.white)
                .cornerRadius(10)
                .fontWeight(.semibold)
            }
            .padding(.horizontal)
            .padding(.vertical, 16)
        }
        .background(Color(.systemBackground))
    }

    private func tagCard(_ tag: Tag) -> some View {
        let isSelected = selectedTagIds.contains(tag.id)

        return Button {
            toggleTag(tag.id)
        } label: {
            VStack(alignment: .leading, spacing: 8) {
                HStack {
                    Circle()
                        .fill(tag.color)
                        .frame(width: 12, height: 12)

                    Text(tag.name)
                        .font(.body)
                        .fontWeight(.bold)
                        .foregroundColor(isSelected ? tag.color : .primary)
                        .lineLimit(1)

                    Spacer()

                    if isSelected {
                        Image(systemName: "checkmark")
                            .foregroundColor(tag.color)
                            .font(.caption)
                    }
                }

                Text("\(tag.count) items")
                    .font(.caption)
                    .foregroundColor(isSelected ? tag.color.opacity(0.7) : .secondary)
            }
            .padding()
            .background(isSelected ? tag.color.opacity(0.1) : Color(.systemGray6))
            .cornerRadius(12)
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .stroke(isSelected ? tag.color : Color.clear, lineWidth: 2)
            )
        }
        .buttonStyle(.plain)
    }

    private func tagListItem(_ tag: Tag, rank: Int) -> some View {
        let isSelected = selectedTagIds.contains(tag.id)

        return HStack {
            Circle()
                .fill(tag.color.opacity(0.2))
                .frame(width: 40, height: 40)
                .overlay {
                    Text("\(rank)")
                        .font(.headline)
                        .fontWeight(.bold)
                        .foregroundColor(tag.color)
                }

            VStack(alignment: .leading, spacing: 4) {
                Text(tag.name)
                    .font(.body)
                    .fontWeight(.bold)
                    .foregroundColor(isSelected ? tag.color : .primary)

                Text("\(tag.count) items • \(tag.category.uppercased())")
                    .font(.caption)
                    .foregroundColor(isSelected ? tag.color.opacity(0.7) : .secondary)
            }

            Spacer()

            if tag.isPopular {
                Image(systemName: "chart.line.uptrend.xyaxis")
                    .foregroundColor(.orange)
                    .font(.caption)
            }

            Button {
                toggleTag(tag.id)
            } label: {
                Image(systemName: isSelected ? "checkmark.circle.fill" : "circle")
                    .foregroundColor(isSelected ? tag.color : .gray)
            }
            .buttonStyle(.plain)
        }
        .contentShape(Rectangle())
        .onTapGesture {
            toggleTag(tag.id)
        }
    }

    private func selectedTagChip(_ tag: Tag) -> some View {
        HStack {
            Circle()
                .fill(Color.white.opacity(0.3))
                .frame(width: 20, height: 20)
                .overlay {
                    Text(tag.count > 1000 ? "\(tag.count / 1000)k" : "\(tag.count)")
                        .font(.caption2)
                        .fontWeight(.bold)
                        .foregroundColor(.white)
                }

            Text(tag.name)
                .font(.body)
                .fontWeight(.medium)
                .foregroundColor(.white)

            Spacer()

            Button {
                toggleTag(tag.id)
            } label: {
                Image(systemName: "xmark")
                    .foregroundColor(.white)
                    .font(.caption)
            }
        }
        .padding()
        .background(tag.color)
        .cornerRadius(12)
    }

    private func loadTags() {
        isLoading = true

        // Simulate loading tags
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.8) {
            allTags = generateMockTags()
            filteredTags = allTags

            // Load initial selected tags
            if let selectedTags = selectedTags {
                selectedTagIds = Set(allTags.filter { selectedTags.contains($0.name) }.map { $0.id })
            }

            isLoading = false
        }
    }

    private func generateMockTags() -> [Tag] {
        let tagData: [(String, String, Color, Int)] = [
            ("Flutter", "technology", .blue, 2450),
            ("SwiftUI", "technology", .blue, 1820),
            ("React", "technology", .cyan, 3200),
            ("JavaScript", "technology", .yellow, 4150),
            ("Python", "technology", .green, 3850),
            ("UI Design", "design", .red, 2100),
            ("UX Research", "design", .red, 1550),
            ("Figma", "design", .orange, 1890),
            ("Adobe XD", "design", .purple, 920),
            ("Prototyping", "design", .red, 1340),
            ("Startup", "business", .mint, 2800),
            ("Entrepreneurship", "business", .mint, 1950),
            ("Marketing", "business", .mint, 2350),
            ("SEO", "marketing", .indigo, 1780),
            ("Content Marketing", "marketing", .indigo, 1420),
            ("Social Media", "marketing", .indigo, 2680),
            ("Machine Learning", "science", .green, 3100),
            ("Data Science", "science", .green, 2750),
            ("AI", "science", .green, 3450),
            ("Blockchain", "technology", .yellow, 1650),
            ("Cryptocurrency", "technology", .yellow, 1890),
            ("Web Development", "development", .pink, 3800),
            ("Mobile Development", "development", .pink, 2650),
            ("DevOps", "development", .pink, 1950),
            ("Cloud Computing", "technology", .blue, 2450),
            ("AWS", "technology", .orange, 1780),
            ("Docker", "development", .blue, 1650),
            ("Kubernetes", "development", .blue, 1420),
            ("Photography", "lifestyle", .pink, 1890),
            ("Travel", "lifestyle", .yellow, 2340),
        ]

        return tagData.enumerated().map { index, data in
            Tag(
                id: "tag_\(index)",
                name: data.0,
                color: data.2,
                count: data.3,
                category: data.1,
                isPopular: data.3 > 2000
            )
        }
    }

    private func filterTags() {
        filteredTags = allTags.filter { tag in
            let matchesSearch = searchText.isEmpty || tag.name.localizedCaseInsensitiveContains(searchText)
            let matchesCategory = selectedCategory == "all" || tag.category == selectedCategory
            return matchesSearch && matchesCategory
        }

        sortTags()
    }

    private func sortTags() {
        switch sortBy {
        case "alphabetical":
            filteredTags.sort { $0.name < $1.name }
        case "popularity":
            filteredTags.sort { $0.count > $1.count }
        case "recent":
            // In a real app, you'd sort by creation/update time
            filteredTags.sort { $0.count > $1.count }
        default:
            break
        }
    }

    private func toggleTag(_ tagId: String) {
        if selectedTagIds.contains(tagId) {
            selectedTagIds.remove(tagId)
        } else {
            selectedTagIds.insert(tagId)
        }
    }

    private func applyTags() {
        let selectedTagNames = allTags
            .filter { selectedTagIds.contains($0.id) }
            .map { $0.name }

        // Return selected tag names to parent view
        dismiss()
    }
}

#Preview {
    TagsScreen()
}

File: snippets/screens/flutter/tags_screen/3/tags_screen.dart
import 'package:flutter/material.dart';

class Tag {
  final String id;
  final String name;
  final String? color;
  final int count;
  final String? category;
  final bool isPopular;

  Tag({
    required this.id,
    required this.name,
    this.color,
    required this.count,
    this.category,
    this.isPopular = false,
  });
}

class TagsScreen extends StatefulWidget {
  final List<String>? selectedTags;

  const TagsScreen({super.key, this.selectedTags});

  @override
  State<TagsScreen> createState() => _TagsScreenState();
}

class _TagsScreenState extends State<TagsScreen> with TickerProviderStateMixin {
  late TabController _tabController;
  final TextEditingController _searchController = TextEditingController();

  List<Tag> _allTags = [];
  List<Tag> _filteredTags = [];
  Set<String> _selectedTags = {};
  String _selectedCategory = 'all';
  String _sortBy = 'popularity';
  bool _isLoading = true;

  final List<String> _categories = [
    'all',
    'technology',
    'design',
    'business',
    'development',
    'marketing',
    'science',
    'education',
    'health',
    'lifestyle',
  ];

  @override
  void initState() {
    super.initState();
    _tabController = TabController(length: 3, vsync: this);
    _selectedTags = Set<String>.from(widget.selectedTags ?? []);
    _searchController.addListener(_filterTags);
    _loadTags();
  }

  @override
  void dispose() {
    _tabController.dispose();
    _searchController.dispose();
    super.dispose();
  }

  Future<void> _loadTags() async {
    setState(() => _isLoading = true);

    // Simulate API call
    await Future.delayed(const Duration(milliseconds: 800));

    final mockTags = _generateMockTags();

    setState(() {
      _allTags = mockTags;
      _filteredTags = mockTags;
      _isLoading = false;
    });
  }

  List<Tag> _generateMockTags() {
    final tagData = [
      {'name': 'Flutter', 'category': 'technology', 'count': 2450, 'color': '#0175C2'},
      {'name': 'Dart', 'category': 'technology', 'count': 1820, 'color': '#0175C2'},
      {'name': 'React', 'category': 'technology', 'count': 3200, 'color': '#61DAFB'},
      {'name': 'JavaScript', 'category': 'technology', 'count': 4150, 'color': '#F7DF1E'},
      {'name': 'Python', 'category': 'technology', 'count': 3850, 'color': '#3776AB'},
      {'name': 'UI Design', 'category': 'design', 'count': 2100, 'color': '#FF6B6B'},
      {'name': 'UX Research', 'category': 'design', 'count': 1550, 'color': '#FF6B6B'},
      {'name': 'Figma', 'category': 'design', 'count': 1890, 'color': '#F24E1E'},
      {'name': 'Adobe XD', 'category': 'design', 'count': 920, 'color': '#FF61F6'},
      {'name': 'Prototyping', 'category': 'design', 'count': 1340, 'color': '#FF6B6B'},
      {'name': 'Startup', 'category': 'business', 'count': 2800, 'color': '#4ECDC4'},
      {'name': 'Entrepreneurship', 'category': 'business', 'count': 1950, 'color': '#4ECDC4'},
      {'name': 'Marketing', 'category': 'business', 'count': 2350, 'color': '#4ECDC4'},
      {'name': 'SEO', 'category': 'marketing', 'count': 1780, 'color': '#45B7D1'},
      {'name': 'Content Marketing', 'category': 'marketing', 'count': 1420, 'color': '#45B7D1'},
      {'name': 'Social Media', 'category': 'marketing', 'count': 2680, 'color': '#45B7D1'},
      {'name': 'Machine Learning', 'category': 'science', 'count': 3100, 'color': '#96CEB4'},
      {'name': 'Data Science', 'category': 'science', 'count': 2750, 'color': '#96CEB4'},
      {'name': 'AI', 'category': 'science', 'count': 3450, 'color': '#96CEB4'},
      {'name': 'Blockchain', 'category': 'technology', 'count': 1650, 'color': '#FECA57'},
      {'name': 'Cryptocurrency', 'category': 'technology', 'count': 1890, 'color': '#FECA57'},
      {'name': 'Web Development', 'category': 'development', 'count': 3800, 'color': '#FF9FF3'},
      {'name': 'Mobile Development', 'category': 'development', 'count': 2650, 'color': '#FF9FF3'},
      {'name': 'DevOps', 'category': 'development', 'count': 1950, 'color': '#FF9FF3'},
      {'name': 'Cloud Computing', 'category': 'technology', 'count': 2450, 'color': '#54A0FF'},
      {'name': 'AWS', 'category': 'technology', 'count': 1780, 'color': '#FF9500'},
      {'name': 'Docker', 'category': 'development', 'count': 1650, 'color': '#0db7ed'},
      {'name': 'Kubernetes', 'category': 'development', 'count': 1420, 'color': '#326ce5'},
      {'name': 'Photography', 'category': 'lifestyle', 'count': 1890, 'color': '#FD79A8'},
      {'name': 'Travel', 'category': 'lifestyle', 'count': 2340, 'color': '#FDCB6E'},
    ];

    return tagData.asMap().entries.map((entry) {
      final index = entry.key;
      final data = entry.value;
      return Tag(
        id: 'tag_$index',
        name: data['name'] as String,
        category: data['category'] as String,
        count: data['count'] as int,
        color: data['color'] as String?,
        isPopular: (data['count'] as int) > 2000,
      );
    }).toList();
  }

  void _filterTags() {
    final query = _searchController.text.toLowerCase();
    setState(() {
      _filteredTags = _allTags.where((tag) {
        final matchesSearch = tag.name.toLowerCase().contains(query);
        final matchesCategory = _selectedCategory == 'all' ||
                               tag.category == _selectedCategory;
        return matchesSearch && matchesCategory;
      }).toList();

      _sortTags();
    });
  }

  void _sortTags() {
    switch (_sortBy) {
      case 'alphabetical':
        _filteredTags.sort((a, b) => a.name.compareTo(b.name));
        break;
      case 'popularity':
        _filteredTags.sort((a, b) => b.count.compareTo(a.count));
        break;
      case 'recent':
        // In a real app, you'd sort by creation/update time
        _filteredTags.sort((a, b) => b.count.compareTo(a.count));
        break;
    }
  }

  void _toggleTag(String tagId) {
    setState(() {
      if (_selectedTags.contains(tagId)) {
        _selectedTags.remove(tagId);
      } else {
        _selectedTags.add(tagId);
      }
    });
  }

  void _applyTags() {
    final selectedTagNames = _allTags
        .where((tag) => _selectedTags.contains(tag.id))
        .map((tag) => tag.name)
        .toList();

    Navigator.pop(context, selectedTagNames);
  }

  Color _getTagColor(String? colorHex) {
    if (colorHex == null) return Colors.blue;
    final hex = colorHex.replaceAll('#', '');
    return Color(int.parse('FF$hex', radix: 16));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Tags (${_selectedTags.length} selected)'),
        actions: [
          if (_selectedTags.isNotEmpty)
            TextButton(
              onPressed: _applyTags,
              child: const Text('Apply'),
            ),
        ],
        bottom: TabBar(
          controller: _tabController,
          tabs: const [
            Tab(text: 'Browse', icon: Icon(Icons.explore)),
            Tab(text: 'Popular', icon: Icon(Icons.trending_up)),
            Tab(text: 'My Tags', icon: Icon(Icons.bookmark)),
          ],
        ),
      ),
      body: Column(
        children: [
          // Search Bar
          Container(
            padding: const EdgeInsets.all(16),
            child: TextField(
              controller: _searchController,
              decoration: InputDecoration(
                hintText: 'Search tags...',
                prefixIcon: const Icon(Icons.search),
                suffixIcon: _searchController.text.isNotEmpty
                    ? IconButton(
                        icon: const Icon(Icons.clear),
                        onPressed: () {
                          _searchController.clear();
                          _filterTags();
                        },
                      )
                    : null,
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(12),
                ),
                filled: true,
                fillColor: Theme.of(context).colorScheme.surface,
              ),
            ),
          ),

          // Filters and Sort
          Container(
            padding: const EdgeInsets.symmetric(horizontal: 16),
            child: Row(
              children: [
                // Category Filter
                Expanded(
                  child: DropdownButtonFormField<String>(
                    value: _selectedCategory,
                    decoration: const InputDecoration(
                      labelText: 'Category',
                      border: OutlineInputBorder(),
                      isDense: true,
                    ),
                    items: _categories.map((category) {
                      return DropdownMenuItem(
                        value: category,
                        child: Text(category == 'all'
                            ? 'All Categories'
                            : category.toUpperCase()),
                      );
                    }).toList(),
                    onChanged: (value) {
                      setState(() {
                        _selectedCategory = value!;
                        _filterTags();
                      });
                    },
                  ),
                ),

                const SizedBox(width: 16),

                // Sort Dropdown
                Expanded(
                  child: DropdownButtonFormField<String>(
                    value: _sortBy,
                    decoration: const InputDecoration(
                      labelText: 'Sort by',
                      border: OutlineInputBorder(),
                      isDense: true,
                    ),
                    items: const [
                      DropdownMenuItem(
                        value: 'popularity',
                        child: Text('Most Popular'),
                      ),
                      DropdownMenuItem(
                        value: 'alphabetical',
                        child: Text('A-Z'),
                      ),
                      DropdownMenuItem(
                        value: 'recent',
                        child: Text('Recently Added'),
                      ),
                    ],
                    onChanged: (value) {
                      setState(() {
                        _sortBy = value!;
                        _sortTags();
                      });
                    },
                  ),
                ),
              ],
            ),
          ),

          const SizedBox(height: 16),

          // Content
          Expanded(
            child: TabBarView(
              controller: _tabController,
              children: [
                _buildBrowseTab(),
                _buildPopularTab(),
                _buildMyTagsTab(),
              ],
            ),
          ),
        ],
      ),
      bottomNavigationBar: _selectedTags.isNotEmpty
          ? Container(
              padding: const EdgeInsets.all(16),
              decoration: BoxDecoration(
                color: Theme.of(context).scaffoldBackgroundColor,
                boxShadow: [
                  BoxShadow(
                    color: Colors.black.withOpacity(0.1),
                    blurRadius: 4,
                    offset: const Offset(0, -2),
                  ),
                ],
              ),
              child: SafeArea(
                child: Row(
                  children: [
                    Expanded(
                      child: OutlinedButton(
                        onPressed: () {
                          setState(() => _selectedTags.clear());
                        },
                        child: const Text('Clear All'),
                      ),
                    ),
                    const SizedBox(width: 16),
                    Expanded(
                      flex: 2,
                      child: ElevatedButton(
                        onPressed: _applyTags,
                        child: Text('Apply ${_selectedTags.length} Tags'),
                      ),
                    ),
                  ],
                ),
              ),
            )
          : null,
    );
  }

  Widget _buildBrowseTab() {
    if (_isLoading) {
      return const Center(child: CircularProgressIndicator());
    }

    if (_filteredTags.isEmpty) {
      return Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(Icons.search_off, size: 64, color: Colors.grey[400]),
            const SizedBox(height: 16),
            Text(
              'No tags found',
              style: Theme.of(context).textTheme.titleLarge,
            ),
            const SizedBox(height: 8),
            Text(
              'Try adjusting your search or filters',
              style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                color: Colors.grey[600],
              ),
            ),
          ],
        ),
      );
    }

    return GridView.builder(
      padding: const EdgeInsets.all(16),
      gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
        crossAxisCount: 2,
        childAspectRatio: 3,
        crossAxisSpacing: 12,
        mainAxisSpacing: 12,
      ),
      itemCount: _filteredTags.length,
      itemBuilder: (context, index) {
        final tag = _filteredTags[index];
        return _buildTagCard(tag);
      },
    );
  }

  Widget _buildPopularTab() {
    final popularTags = _allTags.where((tag) => tag.isPopular).toList();
    popularTags.sort((a, b) => b.count.compareTo(a.count));

    return ListView.builder(
      padding: const EdgeInsets.all(16),
      itemCount: popularTags.length,
      itemBuilder: (context, index) {
        final tag = popularTags[index];
        return _buildTagListItem(tag, index + 1);
      },
    );
  }

  Widget _buildMyTagsTab() {
    final myTags = _allTags.where((tag) => _selectedTags.contains(tag.id)).toList();

    if (myTags.isEmpty) {
      return Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(Icons.bookmarks, size: 64, color: Colors.grey[400]),
            const SizedBox(height: 16),
            Text(
              'No tags selected',
              style: Theme.of(context).textTheme.titleLarge,
            ),
            const SizedBox(height: 8),
            Text(
              'Browse tags and add them to your collection',
              style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                color: Colors.grey[600],
              ),
            ),
            const SizedBox(height: 24),
            ElevatedButton(
              onPressed: () => _tabController.animateTo(0),
              child: const Text('Browse Tags'),
            ),
          ],
        ),
      );
    }

    return Padding(
      padding: const EdgeInsets.all(16),
      child: Wrap(
        spacing: 8,
        runSpacing: 8,
        children: myTags.map((tag) => _buildSelectedTagChip(tag)).toList(),
      ),
    );
  }

  Widget _buildTagCard(Tag tag) {
    final isSelected = _selectedTags.contains(tag.id);
    final tagColor = _getTagColor(tag.color);

    return Card(
      elevation: isSelected ? 4 : 1,
      color: isSelected ? tagColor.withOpacity(0.1) : null,
      child: InkWell(
        onTap: () => _toggleTag(tag.id),
        borderRadius: BorderRadius.circular(12),
        child: Container(
          padding: const EdgeInsets.all(12),
          decoration: BoxDecoration(
            borderRadius: BorderRadius.circular(12),
            border: isSelected
                ? Border.all(color: tagColor, width: 2)
                : null,
          ),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Row(
                children: [
                  Container(
                    width: 12,
                    height: 12,
                    decoration: BoxDecoration(
                      color: tagColor,
                      shape: BoxShape.circle,
                    ),
                  ),
                  const SizedBox(width: 8),
                  Expanded(
                    child: Text(
                      tag.name,
                      style: TextStyle(
                        fontWeight: FontWeight.bold,
                        color: isSelected ? tagColor : null,
                      ),
                      maxLines: 1,
                      overflow: TextOverflow.ellipsis,
                    ),
                  ),
                  if (isSelected) Icon(Icons.check, color: tagColor, size: 16),
                ],
              ),
              const SizedBox(height: 4),
              Align(
                alignment: Alignment.centerLeft,
                child: Text(
                  '${tag.count} items',
                  style: Theme.of(context).textTheme.bodySmall?.copyWith(
                    color: isSelected
                        ? tagColor.withOpacity(0.7)
                        : Colors.grey[600],
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildTagListItem(Tag tag, int rank) {
    final isSelected = _selectedTags.contains(tag.id);
    final tagColor = _getTagColor(tag.color);

    return Card(
      margin: const EdgeInsets.only(bottom: 8),
      child: ListTile(
        leading: CircleAvatar(
          backgroundColor: tagColor.withOpacity(0.2),
          child: Text(
            '$rank',
            style: TextStyle(
              color: tagColor,
              fontWeight: FontWeight.bold,
            ),
          ),
        ),
        title: Text(
          tag.name,
          style: TextStyle(
            fontWeight: FontWeight.bold,
            color: isSelected ? tagColor : null,
          ),
        ),
        subtitle: Text(
          '${tag.count} items • ${tag.category?.toUpperCase()}',
          style: TextStyle(
            color: isSelected
                ? tagColor.withOpacity(0.7)
                : Colors.grey[600],
          ),
        ),
        trailing: Row(
          mainAxisSize: MainAxisSize.min,
          children: [
            if (tag.isPopular)
              Icon(Icons.trending_up, color: Colors.orange[600], size: 20),
            const SizedBox(width: 8),
            Checkbox(
              value: isSelected,
              onChanged: (_) => _toggleTag(tag.id),
              activeColor: tagColor,
            ),
          ],
        ),
        onTap: () => _toggleTag(tag.id),
        selected: isSelected,
      ),
    );
  }

  Widget _buildSelectedTagChip(Tag tag) {
    final tagColor = _getTagColor(tag.color);

    return Chip(
      label: Text(
        tag.name,
        style: const TextStyle(color: Colors.white),
      ),
      backgroundColor: tagColor,
      deleteIcon: const Icon(Icons.close, color: Colors.white, size: 18),
      onDeleted: () => _toggleTag(tag.id),
      avatar: CircleAvatar(
        backgroundColor: Colors.white.withOpacity(0.3),
        child: Text(
          tag.count > 1000
              ? '${(tag.count / 1000).toStringAsFixed(0)}k'
              : tag.count.toString(),
          style: const TextStyle(
            color: Colors.white,
            fontSize: 10,
            fontWeight: FontWeight.bold,
          ),
        ),
      ),
    );
  }
}